src/Collections/Sequences/Sorting.dfy

```diff
@@ -0,0 +1,173 @@
+    LexicographicByteSeqBelowAux(x, y, 0)
+  }
+
+  predicate method LexicographicByteSeqBelowAux(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    decreases |x| - n
+  {
+    || n == |x|
+    || (n != |y| && x[n] < y[n])
+    || (n != |y| && x[n] == y[n] && LexicographicByteSeqBelowAux(x, y, n + 1))
+  }
+
+  lemma AboutLexicographicByteSeqBelow()
+    ensures TotalOrdering(LexicographicByteSeqBelow)
+  {
+    assert Reflexive(LexicographicByteSeqBelow) by {
+      forall x, n | 0 <= n <= |x| {
+        AboutLexicographicByteSeqBelowAux_Reflexive(x, n);
+      }
+    }
+    assert AntiSymmetric(LexicographicByteSeqBelow) by {
+      forall x, y, n: nat |
+        n <= |x| && n <= |y| && x[..n] == y[..n] &&
+        LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, x, n)
+      {
+        AboutLexicographicByteSeqBelowAux_AntiSymmetric(x, y, n);
+      }
+    }
+    assert StandardLibrary.Transitive(LexicographicByteSeqBelow) by {
+      forall x, y, z, n: nat |
+        n <= |x| && n <= |y| && n <= |z| &&
+        LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, z, n)
+      {
+        AboutLexicographicByteSeqBelowAux_Transitive(x, y, z, n);
+      }
+    }
+    assert Connected(LexicographicByteSeqBelow) by {
+      forall x, y, n: nat | n <= |x| && n <= |y| {
+        AboutLexicographicByteSeqBelowAux_Connected(x, y, n);
+      }
+    }
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Reflexive(x: seq<uint8>, n: nat)
+    requires n <= |x|
+    ensures LexicographicByteSeqBelowAux(x, x, n)
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_AntiSymmetric(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    requires x[..n] == y[..n]
+    requires LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, x, n)
+    ensures x == y
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Transitive(x: seq<uint8>, y: seq<uint8>, z: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y| && n <= |z|
+    requires LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, z, n)
+    ensures LexicographicByteSeqBelowAux(x, z, n)
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Connected(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    ensures LexicographicByteSeqBelowAux(x, y, n) || LexicographicByteSeqBelowAux(y, x, n)
+    decreases |x| - n
+  {
+  }
+
+  predicate method SortedBy<T>(a: seq<T>,compare: (T, T) -> bool ) {
+    forall i, j | 0 <= i < j < |a| :: compare(a[i],a[j])
+  }
+
+  lemma LemmaNewFirstElementStillSortedBy<T>(x: T, s: seq<T>,  compare: (T, T) -> bool) 
+    requires SortedBy(s, compare)
+    requires |s| == 0 || compare(x,s[0])
+  {}
+
+function method MergeSortBy<T>(a: seq<T>, compare: (T, T) -> bool ): (result :seq<T>)
+    ensures multiset(a) == multiset(result)
```

Comment:
You also probably need a postcondition that the result is sorted.

Reply:
In order to add `SortedBy` in a postcondition here, you'll need to export it from the module, as well.

Reply:
And I think you'll also need these preconditions:

```dafny
requires StandardLibrary.Transitive(compare)
requires Connected(compare)
```